### ISSUE TITLE
Best way to add a time out?

### ISSUE THREAD

**Issue by nah0y** [Real-Serious-Games/C-Sharp-Promise#60]
Hey there :)

I don't think there is a builtin thing to add a time out to a promise, so how would you do it?
The only thing I can think of right now is after I start my promises, I just do something like TimeOutHelper.RegisterPromise(promise) that will start a coroutine, wait for x seconds, and then reject the coroutine if it's not done.

Thanks!

**Comment by RoryDungan** [comment]
Would something like this work?
```cs
IPromise ResolveOrTimeout(IPromise promise, float timeout)
{
    var newPromise = new Promise();
    var resolvedBeforeTimeout = false;

    PromiseTimer.WaitFor(timeout)
        .Then(() => {
            if (!resolvedBeforeTimeout)
                newPromise.Reject(new Exception("Promsie timed out"));
            }
        });

    promise
        .Then(() => {
            resolvedBeforeTimeout = true;
            newPromise.Resolve();
        })
        .Catch(ex => {
            resolvedBeforeTimeout = true;
            newPromsie.Reject(ex));
        });

    return newPromise;
}
```

**Comment by ashleydavis** [comment]
Check out my article...

http://www.what-could-possibly-go-wrong.com/promises-for-game-development/

There's an example of how to do time outs with the race function.

**Comment by RoryDungan** [comment]
That was what I thought of first too and would be much simpler, but I think @nah0y wanted the promise to reject if it times out.

**Comment by ashleydavis** [comment]
In that case make your Timeout function reject instead of resolve, that should reject the raced promise (unless my memory is incorrect - I can't fully remember how race works).

**Comment by nah0y** [comment]
Thanks for all your answers and sorry for the time it took to answer...

I find it way easier and using less line of codes doing just this:

`
void Start ()
{
        Promise myPromise = new Promise();
        StartCoroutine(TimeOutPromiseCoroutine(myPromise, 2f));
}

private IEnumerator TimeOutPromiseCoroutine(Promise promise, float timeOut)
    {
        yield return new WaitForSeconds(timeOut);

        if (promise.CurState == PromiseState.Pending)
        {
			promise.Reject(new Exception("Promise has timed out"));
        }
    }
`

I can reuse the coroutine everywhere I want, instead of adding the same code before starting any promise (same thing @ashleydavis suggest in his blog post).
And also, I prefer not to use the PromiseTimer, because I have to create an instance as a class variable, and call Update on it in an update method. And if I want to use multiple promise timers, each of them needs to be added in the Update method etc... kind of "hard" to manipulate.
I think it's more elegant with a simple coroutine.
But maybe I'm wrong?

**Comment by RoryDungan** [comment]
Looks like a neat solution! With the PromiseTimer, the pattern that I have found generally works best is to just have a single instance of it that is shared throughout your app. An easy way to manage this is to create a [singleton object](https://wiki.unity3d.com/index.php/Singleton) which updates the PromiseTimer and then use that everywhere else in your code. 

The only case where I might use multiple PromiseTimers would be maybe if I had some things that I wanted updated on FixedUpdate or at a different rate to the regular Update loop.

**Comment by nah0y** [comment]
"the pattern that I have found generally works best is to just have a single instance of it that is shared throughout your app"
I don't understand how you can share it. Once a promise timer is consumed, you can't reuse it right?

**Comment by RoryDungan** [comment]
You can call WaitFor, WaitUntil or WaitWhile on the same PromiseTimer object as many times as you want and each time will return a new promise. The starting time is the time at which you called WaitFor, not the time the PromiseTimer was constructed.

**Comment by nah0y** [comment]
Oh I didn't know. You mean you could do PromiseTimer.WaitFor directly?
Without having to do use an instance like this?

```
PromiseTimer myTimer = new PromiseTimer();
myTimer.WaitFor
```

If that's the case, I maybe didn't have the latest version when I tested.

**Comment by ashleydavis** [comment]
Instance once and use as many times as you like.

PromiseTimer myTimer = new PromiseTimer();
myTimer.WaitFor(....).Then(something);
myTimer.WaitFor(...).Then(somethingElse);
myTimer.WaitFor(...).Then(keepGoingAsManyTimesAsYouWouldLike);

**Comment by nah0y** [comment]
Thanks for the answer :)
I didn't know we could do that (would be nice to have that in the documentation :D).

This means every time I call WaitFor, it "reset" the state of the promise.

**Comment by ashleydavis** [comment]
It creates a _separate_ promise every time you call WaitFor.

**Comment by adamsingle** [comment]
I put together a quick example of a singleton like Ash mentioned above. You can find it [here](https://gist.github.com/adamsingle/d0010912e23f89b6ed473e868b65cdcc)  To use it you would just call

Promise waitForPromiseOne = PromiseTimerManager.Instance.WaitFor(secondsToWait);

waitForPromiseOne.Then(() => /*do something*/);

calling this function again gives you a new promise that is independent of this first one

Promise waitForPromiseTwo = PromiseTimerManager.Instance.WaitFor(secondsToWait);

waitForPromiseTwo.Then(() => /*do something else*/);

This gives you a single place, and a single promise timer, that can handle all your promise timer needs instead of each class needing their own.

I hope this helps.

**Comment by FNGgames** [comment]
Sorry to necro the thread, just found it and attempted to build something by putting together @ashleydavis's answer with @adamsingle's manager setup. 

I've gone for an extension method because it preserves the fluid-API of the original library and lets you chain a timeout onto any promise like so:

`ExamplePromise().Timeout(5).Done(OnPromiseResolved, OnPromiseRejected);`

Where the `Timeout` method is esentially calling @ashleydavis's code in the first answer:

```csharp
public static class Promises
{
    private static readonly PromiseTimer _timer = new();

    public static void Update(float deltaTime) => _timer.Update(deltaTime);

    public static IPromise WaitFor(float seconds) => _timer.WaitFor(seconds);

    public static IPromise WaitUntil(Func<TimeData, bool> predicate) => _timer.WaitUntil(predicate);

    public static IPromise WaitWhile(Func<TimeData, bool> predicate) => _timer.WaitWhile(predicate);

    public static IPromise Timeout(this IPromise initialPromise, float timeout)
    {
        var promise = new Promise();
        var resolved = false;

        WaitFor(timeout).Then(() =>
        {
            if (!resolved) promise.Reject(new PromiseTimedOutException(timeout));
        });

        initialPromise
            .Then(() =>
            {
                resolved = true;
                promise.Resolve();
            })
            .Catch(exception => 
            {
                resolved = true;
                promise.Reject(exception);
            });

        return promise;
    }
}
```

The whole setup with an example script is in [this gist](https://gist.github.com/FNGgames/8bfae62793dc860514f61f05ec35a50b).
